lib/Net/BGP/Transport.pm

```diff
@@ -656,8 +659,8 @@ sub _handle_socket_read_ready
     my $socket = $this->{_peer_socket};
 
     unless (defined $socket) {
-      warn $this->parent->asstring . ": Connection lost - Connection is formaly shutdown now\n";
-      $this->_cease;
+      warn $this->parent->asstring . ": Connection lost - Connection is fully shutdown now\n";
+      $this->_kill_session;
```

Comment:
This is a bit problematic because `_kill_session()` will call the error callback with `undef`, and it expects to receive a `Net::BGP::Notification` object which this peer has sent to its peer.

Since this is special case band-aid code introduced by improper handling of collision detection elsewhere in the code, it's OK for now to special-case it, not calling `_kill_session()` and instead do the following:

1. Call `close_session()`.
2. Invoke the reset callback.

Reply:
You're not wrong, there are definitely a bunch of issues in the collision avoidance code. After suggesting these changes,  I am concerned that they may make it harder to actually track down those issues.

The problem seems to occur most often when you have two Net::BGP sessions talking to each other. This seems to imply that Net::BGP is not following the best practice procotol and is relying on something else (i.e. a router) doing the right thing.

Reply:
How would you like to proceed then? Do you think these fixes are still worth pursuing at this time? It's been a very long time since I've actually used Net::BGP myself, so I have little visibility into the severity of the issues or how frequently they occur in day-to-day use.

A few years back I started refactoring Net::BGP to use the CPAN IO::Async module for its I/O needs, allowing it to use a standardized event loop and I/O API and making it easier to embed Net::BGP along with other services such as an HTTP server. Since the changes are quite fundamental and destructive, perhaps it would offer a better opportunity to undertake a serious cleanup of the collision avoidance code.

Reply:
I agree that refactoring it to use IO::Async would make a lot of sense. How far did you get, would you be able to complete it to a point where I could test it for you? I too have somewhat moved on, but every so often I do come up with a need for Net::BGP, but have discarded the idea feeling that the code had somewhat been abandoned.

If you are back online so to speak I could definitely devote some resourced in to getting this whole thing working.

For the moment stick the changes in, they are good defensive code anyway (though I might put some debug in there that essentially says "this is broken!" so that it can be tracked).

Reply:
It's a first-cut refactor, with naive mapping from the IO::Async API to existing Net::BGP handler routines. I'm under no illusions that the coverage is comprehensive, but all the existing tests pass.

I doubt that these changes in of themselves will fix any of the collision avoidance issues. For that, more work would need to be done to better define the interfaces between the IO::Async entry-points and the Transport.pm event handling machinery. But it might still be interesting to see what issues there are that the test coverage doesn't address.

It's in the [IO-Async-Refactor](https://github.com/singularsyntax/CPAN-Net-BGP/tree/IO-Async-Refactor) branch, if you have some time to give it a spin.